services/worker/src/worker/job_runners/config/parquet_and_info.py

```diff
@@ -643,37 +650,58 @@ def fill_builder_info(
     builder.info.download_size = 0
     builder.info.dataset_size = 0
     logging.info("Start validation of parquet files.")
-    for split in data_files:
+    num_examples = 0
+    for split, urls in data_files.items():
         split = str(split)  # in case it's a NamedSplit
+        first_url = urls[0]
         try:
-            parquet_files_and_sizes: list[tuple[pq.ParquetFile, int]] = thread_map(
-                functools.partial(
-                    retry_and_validate_get_parquet_file_and_size,
-                    hf_endpoint=hf_endpoint,
-                    hf_token=hf_token,
-                    validate=validate,
-                ),
-                data_files[split],
-                unit="pq",
-                disable=True,
+            # try to read first file metadata to infer features schema
+            first_pf, first_pf_size = retry_and_validate_get_parquet_file_or_num_examples_and_size(
+                first_url,
+                hf_endpoint,
+                hf_token,
+                validate,
+                return_pf=True,
             )
-            parquet_files, sizes = zip(*parquet_files_and_sizes)
-            logging.info(f"{len(parquet_files)} parquet files are valid for copy. ")
+            if builder.info.features is None:
+                builder.info.features = Features.from_arrow_schema(first_pf.schema_arrow)  # type: ignore
```

Comment:
type: ignore here (and below) because of `Item "int" has no attribute "schema_arrow"`. because `retry_and_validate_get_parquet_file_or_num_examples_and_size` now can return either `pq.ParquetFile` or `int` as a first element, which is shitty but i'm too tired to refactor now...

Reply:
actually did my best here [658eb50](https://github.com/huggingface/datasets-server/pull/2564/commits/658eb50aeec48e634c5f772bfed5ba2eb922eb22)